BlueprintUICommonControls/Sources/AttributedLabel.swift

```diff
@@ -43,53 +57,508 @@ public struct AttributedLabel: Element, Hashable {
+        case link
+        /// Detect phone numbers. Tapping the phone number prompts the user to call it.
+        case phoneNumber
+
+        var checkingType: NSTextCheckingResult.CheckingType {
+            switch self {
+            case .date: return .date
+            case .address: return .address
+            case .link: return .link
+            case .phoneNumber: return .phoneNumber
             }
-            label.accessibilityTraits = union
         }
     }
 }
 
+// MARK: View implementation
+
 extension AttributedLabel {
 
-    private final class LabelView: UILabel {
-        var textRectOffset: UIOffset = .zero {
+    private final class LabelView: UILabel, UIGestureRecognizerDelegate {
+        /// The touch handling logic explicitly tracks the active links when touches begin, so if you drag outside
+        /// the link and touch up over another link, it just cancels the tap rather than accidentally opening
+        /// a different link.
+        private var trackingLinks: [Link]?
+
+        private var linkDetectionTypes: Set<LinkDetectionType> = []
+        private var linkAttributes: [NSAttributedString.Key: Any] = [:]
+        private var activeLinkAttributes: [NSAttributedString.Key: Any] = [:]
+
+        private var links: [Link] = [] {
+            didSet {
+                isUserInteractionEnabled = !links.isEmpty
+            }
+        }
+
+        private var textRectOffset: UIOffset = .zero {
             didSet {
                 if oldValue != textRectOffset {
                     setNeedsDisplay()
                 }
             }
         }
 
+        /// The view needs to keep track of the current index for the accessibility rotor.
+        private var accessibilityLinkIndex = -1
+
+        /// These elements need to be retained by the view, and cannot be created inside the
+        /// `accessibilityCustomRotors` getter.
+        private var accessibilityLinks: [LinkAccessibilityElement] = []
+
+        override var accessibilityCustomRotors: [UIAccessibilityCustomRotor]? {
+            set { fatalError() }
+            get {
+                [
+                    UIAccessibilityCustomRotor(systemType: .link) { [weak self] predicate in
+                        guard let self = self else {
+                            return nil
+                        }
+
+                        self.accessibilityLinkIndex += predicate.searchDirection == .next ? 1 : -1
+                        self.accessibilityLinkIndex = min(self.accessibilityLinks.count - 1, self.accessibilityLinkIndex)
+                        self.accessibilityLinkIndex = max(0, self.accessibilityLinkIndex)
+
+                        let link = self.accessibilityLinks[self.accessibilityLinkIndex]
+                        return UIAccessibilityCustomRotorItemResult(targetElement: link, targetRange: nil)
+                    },
+                ]
+            }
+        }
+
+        var linkHandler: LinkHandler?
```

Comment:
Should this be `weak`? Eg, if someone makes a view controller be their link handler; and the vc retains the blueprint view, and thus this view, we'll have a retain cycle.

Reply:
It's a protocol, unless we force it to be a class it can't be weak. Wouldn't this problem exist for any reference type putting itself into its Blueprint's environment though? The cycle exists there without this I think (I could be wrong?)

Reply:
Hmm, that's true; good point... Fair enough then! This seems fine.